### PR TITLE
enable support for Raft consensus algorithm

### PR DESCRIPTION
**Pull request by lichenxi-webank**

### CODE REVIEW

File: libconsensus/raft/Common.h [hunk 1]
```diff
@@ -0,0 +1,275 @@
+/*
+    This file is part of cpp-ethereum.
```
Comment (wheatli) [inline libconsensus/raft/Common.h]: change cpp-ethereum to FISCO-BCOS

File: libconsensus/raft/Common.h [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+    You should have received a copy of the GNU General Public License
+    along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * @file: Common.h
+ * @author: fisco-dev
+ *
+ * @date: 2017
+ * A proof of work algorithm.
```
Comment (wheatli) [inline libconsensus/raft/Common.h]: Plz give out a right version of comments

File: libconsensus/raft/Common.h [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+{
+    TO_LEADER,
+    TO_FOLLOWER,
+    NONE
+};
+
+struct RaftMsgPacket
+{
+    /// the index of the node that send this raft message
+    u256 nodeIdx;
```
Comment (wheatli) [inline libconsensus/raft/Common.h]: do not use u256

File: libconsensus/raft/Common.h [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+    /// the node id of the node that sends this raft message
+    h512 nodeId;
+    /// type of the packet
+    RaftPacketType packetType;
+    /// the data of concrete request
+    bytes data;
+    /// timestamp of receive this raft message
+    u256 timestamp;
```
Comment (wheatli) [inline libconsensus/raft/Common.h]: do not use u256

File: libconsensus/raft/Common.h [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+            _e << dev::eth::errinfo_name("invalid msg format")
+               << dev::eth::BadFieldError(field, toHex(_rlp[field].data().toBytes()));
+            throw;
+        }
+    }
+};
+
+struct RaftVoteReq : public RaftMsg
+{
+    u256 candidate;
```
Comment (wheatli) [inline libconsensus/raft/Common.h]: we should double check all the type of variables of message packet

File: libconsensus/raft/RaftEngine.h [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    bool isMajorityVote(u256 const& _votes) { return _votes >= m_nodeNum - m_f; }
+
+    dev::p2p::P2PMessage::Ptr generateVoteReq();
+    dev::p2p::P2PMessage::Ptr generateHeartbeat();
+
+    void broadcastVoteReq();
+    void broadcastHeartbeat();
+    void broadcastMsg(dev::p2p::P2PMessage::Ptr _data);
+
+    // handle response 处理响应消息
```
Comment (wheatli) [inline libconsensus/raft/RaftEngine.h]: do not allow chinese

File: libconsensus/raft/Common.h (hunk 3 above)
Comment (lichenxi-webank) [reply libconsensus/raft/Common.h]: to be compatible with  function in Sealer & ConsensusBasicEngine class,  use 256 for node idx temporarily.

**Pull request merged**